src/data/EventD2Repository.ts

```diff
@@ -0,0 +1,117 @@
+import { Id } from "domain/entities/Base";
+import { EventMetadata, Stats } from "domain/entities/Event";
+import { EventRepository } from "domain/repositories/EventRepository";
+import logger from "utils/log";
+
+type EventsRequestWithFilter = EventsGetRequest & { fields: string; event: string[] };
+const eventsFields = "event,eventDate,status,orgUnit,program,programStage,dataValues[dataElement,value]";
+
+export class EventD2Repository implements EventRepository {
+    constructor(private api: D2Api) {}
+
+    async getByIds(ids: Id[]): Async<EventMetadata[]> {
+        const dataValues = await getInChunks(ids, async eventIds => {
+            return this.api.events
+                .get({
+                    event: eventIds.join(";"),
+                    fields: eventsFields,
+                    totalPages: true,
+                    pageSize: 1e6,
+                } as EventsRequestWithFilter)
+                .getData()
+                .then(res => {
+                    return res.events.map(event => {
+                        return {
+                            event: event.event,
+                            orgUnit: event.orgUnit,
+                            program: event.program,
+                            eventDate: event.eventDate,
+                            status: event.status,
+                            programStage: event.programStage,
+                            dataValues: event.dataValues.map(dv => ({
+                                dataElement: dv.dataElement,
+                                value: dv.value,
+                            })),
+                        };
+                    });
+                })
+                .catch(() => {
+                    logger.error(`Error getting events: ${eventIds.join(",")}`);
+                    return [];
+                });
+        });
+        return dataValues;
+    }
+
+    async saveAll(events: EventMetadata[]): Promise<Stats> {
+        const eventsIdsToSave = events.map(event => event.event);
+        const stats = await getInChunks<Stats>(eventsIdsToSave, async eventIds => {
+            return this.api.events
+                .get({
+                    event: eventIds.join(";"),
+                    fields: eventsFields,
+                    totalPages: true,
+                    pageSize: 1e6,
+                } as EventsRequestWithFilter)
+                .getData()
+                .then(res => {
+                    const postEvents = eventIds.map(eventId => {
+                        const existingD2Event = res.events.find(d2Event => d2Event.event === eventId);
+                        const event = events.find(event => event.event === eventId);
+                        if (!event) {
+                            throw Error("Cannot find event");
+                        }
+                        return {
+                            ...(existingD2Event || {}),
+                            ...event,
```

Comment:
Caution: the analysis of completeness of properties must be done in `event.dataValues`. Example:

```
{
"lastUpdated": "2023-06-20T10:14:45.103",
"created": "2023-06-16T13:13:03.150",
"dataElement": "EZXB2m47njy",
"value": "200",
"providedElsewhere": false,
"createdByUserInfo": {
"uid": "SyI5Bx6M0Iv",
"firstName": "Adrian",
"surname": "Quintana",
"username": "adrian@eyeseetea.com"
},
"lastUpdatedByUserInfo": {
"uid": "SyI5Bx6M0Iv",
"firstName": "Adrian",
"surname": "Quintana",
"username": "adrian@eyeseetea.com"
}
}
```

I'd say we need to provide `providedElsewhere: boolean`  to the entity so is not lost on re-save.